tests/integration/test_scheduler.py

```diff
@@ -129,3 +129,65 @@ async def mock_shutdown(*a, **k):
     assert last_record.exc_text.startswith("Traceback (most recent call last)")
     assert ("During handling of the above exception, "
             "another exception occurred") not in last_record.exc_text
+
+
```

Comment:
I noticed this test is passing on master, perhaps it shouldn't be?
https://github.com/cylc/cylc-flow/blob/124b2b29e96f041a4f01b2f86d4e8c198ab2b437/tests/integration/test_scheduler.py#L76

Using `pytest -r A` to debug, I think the reason it is passing might be because the task is in runahead at scheduler start-up, but the bug was found by holding a non-runahead task

Reply:
Yes, it seems runahead tasks are not affected by the bug

Reply:
I think I've found a fix

Reply:
Not sure what you mean. All tasks a runahead at startup. The tasks end up in `pre_submit_tasks` as a result of being released from runahead by `one.pool.release_runahead_tasks`?

Reply:
Test fails for me on master:

```
>           assert one.pre_submit_tasks == []
E           AssertionError: assert [<TaskProxy '//1/one'>] == []
E             Left contains one more item: <TaskProxy '//1/one'>
E             Full diff:
E             - []
E             + [<TaskProxy '//1/one'>]
```

Reply:
Sorry if unclear, I'm talking about `test_resume_does_not_release_tasks`

Reply:
Deleting this broken test, logic covered by the new one.

Reply:
This test fails because it is testing for the wrong thing. The task still has the held status, the bug was because the task was being run despite the held status.

Reply:
I have left this test in.

Because it is only checking that the `is_held` state is preserved it did not identify this bug, it needed a more functional-like test to pick up this problem.

Reply:
This diff makes the test fail on master but passes on this branch
```diff
diff --git a/tests/integration/test_scheduler.py b/tests/integration/test_scheduler.py
index 8b7393bf8..ae9c90631 100644
--- a/tests/integration/test_scheduler.py
+++ b/tests/integration/test_scheduler.py
@@ -21,6 +21,7 @@ from typing import Any, Callable
 
 from cylc.flow.exceptions import CylcError
 from cylc.flow.scheduler import Scheduler
+from cylc.flow.task_state import TASK_STATUS_WAITING
 
 Fixture = Any
 
@@ -78,6 +79,10 @@ async def test_resume_does_not_release_tasks(one: Scheduler, run: Callable):
     schd: Scheduler = one
     async with run(schd):
         assert schd.is_paused
+        # Partial main loop iteration:
+        schd.pool.release_runahead_tasks()
+        schd.release_queued_tasks()
+
         itasks = schd.pool.get_all_tasks()
         assert len(itasks) == 1
         itask = itasks[0]
@@ -85,8 +90,13 @@ async def test_resume_does_not_release_tasks(one: Scheduler, run: Callable):
 
         schd.command_hold('*')
         schd.resume_workflow()
+        # Partial main loop iteration:
+        schd.pool.release_runahead_tasks()
+        schd.release_queued_tasks()
+
         assert not schd.is_paused
         assert itask.state.is_held
+        assert itask.state(TASK_STATUS_WAITING)
 
 
 async def test_shutdown_CylcError_log(one: Scheduler, run: Callable):
```

Reply:
This is covered by the new test. Happy to remove this one or leave it as is.

Reply:
Removed.